File: backend/app/schemas/troca_pedido.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime, date
from decimal import Decimal
from enum import Enum


class StatusTrocaPedido(str, Enum):
    aberto = "aberto"
    aguardando_comercial = "aguardando_comercial"
    aguardando_faturamento = "aguardando_faturamento"
    aguardando_financeiro = "aguardando_financeiro"
    aguardando_ti = "aguardando_ti"
    fechado = "fechado"


class TrocaPedidoCreate(BaseModel):
    franquia_id: int
    motivo: str
    motivo_detalhado: str
    nome_vendedor: str
    numero_pedido_cancelar: str
    data_pedido_cancelar: date
    codigo_produto_cancelar: str
    descricao_pedido_cancelar: str
    numero_novo_pedido: str
    codigo_produto_novo: str
    descricao_novo_pedido: str
    status_portal: str
    nome_cliente: str
    cpf: str
    valor_novo_pedido: Decimal
    valor_pago_cliente: Decimal
    anexos: list[str] = []


class TrocaPedidoResponse(BaseModel):
    id: int
    franquia_id: int
    franquia_nome: Optional[str] = None
    motivo: str
    motivo_detalhado: Optional[str] = None
    nome_vendedor: str
    numero_pedido_cancelar: str
    data_pedido_cancelar: date
    codigo_produto_cancelar: str
    descricao_pedido_cancelar: str
    numero_novo_pedido: str
    codigo_produto_novo: str
    descricao_novo_pedido: str
    status_portal: str
    nome_cliente: Optional[str] = None
    cpf: Optional[str] = None
    valor_novo_pedido: Optional[Decimal] = None
    valor_pago_cliente: Optional[Decimal] = None
    status: StatusTrocaPedido
    anexos: list[str] = []
    observacao_comercial: Optional[str] = None
    observacao_faturamento: Optional[str] = None
    justificativa_reprovacao: Optional[str] = None
    destino_reprovacao: Optional[str] = None
    historico_observacoes: list[dict] = []
    criado_em: datetime
    atualizado_em: datetime

    class Config:
        from_attributes = True


class AprovarTrocaRequest(BaseModel):
    observacao: Optional[str] = None
    anexos: list[str] = []
    destino: Optional[str] = None  # comercial | faturamento | ti | concluir


class ReprovarTrocaRequest(BaseModel):
    justificativa: Optional[str] = None
    destino: Optional[str] = None  # comercial | faturamento | ti | franquia


class ReenviarTrocaRequest(BaseModel):
    franquia_id: int
    motivo: str
    motivo_detalhado: str
    nome_vendedor: str
    numero_pedido_cancelar: str
    data_pedido_cancelar: date
    codigo_produto_cancelar: str
    descricao_pedido_cancelar: str
    numero_novo_pedido: str
    codigo_produto_novo: str
    descricao_novo_pedido: str
    status_portal: str
    nome_cliente: str
    cpf: str
    valor_novo_pedido: Decimal
    valor_pago_cliente: Decimal
    anexos: list[str] = []

File: frontend/src/services/troca-pedido.ts
import api from '@/lib/api'

export interface TrocaPedidoData {
  id: number
  franquia_id: number
  franquia_nome: string
  motivo: string
  motivo_detalhado: string | null
  nome_vendedor: string
  numero_pedido_cancelar: string
  data_pedido_cancelar: string
  codigo_produto_cancelar: string
  descricao_pedido_cancelar: string
  numero_novo_pedido: string
  codigo_produto_novo: string
  descricao_novo_pedido: string
  status_portal: string
  nome_cliente: string | null
  cpf: string | null
  valor_novo_pedido: number | null
  valor_pago_cliente: number | null
  status: 'aberto' | 'aguardando_comercial' | 'aguardando_faturamento' | 'aguardando_financeiro' | 'aguardando_ti' | 'fechado'
  anexos: string[]
  observacao_comercial: string | null
  observacao_faturamento: string | null
  justificativa_reprovacao: string | null
  destino_reprovacao: string | null
  historico_observacoes: { area: string; texto: string; tipo: 'aprovacao' | 'reprovacao'; data: string }[]
  criado_em: string
  atualizado_em: string
}

export interface TrocaPedidoCreatePayload {
  franquia_id: number
  motivo: string
  motivo_detalhado: string
  nome_vendedor: string
  numero_pedido_cancelar: string
  data_pedido_cancelar: string
  codigo_produto_cancelar: string
  descricao_pedido_cancelar: string
  numero_novo_pedido: string
  codigo_produto_novo: string
  descricao_novo_pedido: string
  status_portal: string
  nome_cliente: string
  cpf: string
  valor_novo_pedido: number
  valor_pago_cliente: number
  anexos?: string[]
}

export const trocaPedidoService = {
  async listar(status?: string, franquia_id?: number) {
    const params: Record<string, string | number> = {}
    if (status) params.status = status
    if (franquia_id) params.franquia_id = franquia_id
    const { data } = await api.get('/trocas-pedido', { params })
    return data as TrocaPedidoData[]
  },

  async obter(id: number) {
    const { data } = await api.get(`/trocas-pedido/${id}`)
    return data as TrocaPedidoData
  },

  async criar(payload: TrocaPedidoCreatePayload) {
    const { data } = await api.post('/trocas-pedido', payload)
    return data as TrocaPedidoData
  },

  async aprovar(id: number, opts: { observacao?: string; anexos?: string[]; destino?: string } = {}) {
    const { data } = await api.put(`/trocas-pedido/${id}/aprovar`, opts)
    return data as TrocaPedidoData
  },

  async reprovar(id: number, justificativa?: string, destino?: string) {
    const { data } = await api.put(`/trocas-pedido/${id}/reprovar`, { justificativa, destino })
    return data as TrocaPedidoData
  },

  async reenviar(id: number, payload: TrocaPedidoCreatePayload) {
    const { data } = await api.put(`/trocas-pedido/${id}/reenviar`, payload)
    return data as TrocaPedidoData
  },

  async deletar(id: number) {
    await api.delete(`/trocas-pedido/${id}`)
  },
}
